File: dpb-mcp-go/pkg/analyzer/security.go
package analyzer

import (
	"encoding/json"
	"strings"
	"time"

	"github.com/faithfm/php-dependency-mcp/pkg/composer"
	"github.com/faithfm/php-dependency-mcp/pkg/types"
)

// SecurityAuditResult represents security audit output
type SecurityAuditResult struct {
	Vulnerabilities []types.SecurityVulnerability `json:"vulnerabilities"`
	RiskLevel       string                        `json:"riskLevel"`
	Summary         SecuritySummary               `json:"summary"`
}

// SecuritySummary represents security statistics
type SecuritySummary struct {
	Critical int `json:"critical"`
	High     int `json:"high"`
	Medium   int `json:"medium"`
	Low      int `json:"low"`
}

// AuditSecurity audits dependencies for security issues
func AuditSecurity(repoPath string) (string, error) {
	lock, err := composer.ReadComposerLock(repoPath)
	if err != nil {
		return "", err
	}

	vulnerabilities := make([]types.SecurityVulnerability, 0)

	allPackages := append([]types.PackageInfo{}, lock.Packages...)
	if lock.PackagesDev != nil {
		allPackages = append(allPackages, lock.PackagesDev...)
	}

	for _, pkg := range allPackages {
		// Check for dev versions
		if strings.Contains(pkg.Version, "dev") && !strings.Contains(pkg.Version, "dev-") {
			vulnerabilities = append(vulnerabilities, types.SecurityVulnerability{
				Package:        pkg.Name,
				Version:        pkg.Version,
				Severity:       "medium",
				Description:    "Using development version in production",
				Recommendation: "Pin to a stable release version",
			})
		}

		// Check for pre-1.0 versions
		if strings.HasPrefix(pkg.Version, "0.") {
			vulnerabilities = append(vulnerabilities, types.SecurityVulnerability{
				Package:        pkg.Name,
				Version:        pkg.Version,
				Severity:       "low",
				Description:    "Using pre-1.0 version (potentially unstable)",
				Recommendation: "Consider upgrading to a stable 1.x+ version if available",
			})
		}

		// Check for very old packages
		if pkg.Time != "" {
			pkgTime, err := time.Parse(time.RFC3339, pkg.Time)
			if err == nil {
				fiveYearsAgo := time.Now().AddDate(-5, 0, 0)
				if pkgTime.Before(fiveYearsAgo) {
					vulnerabilities = append(vulnerabilities, types.SecurityVulnerability{
						Package:        pkg.Name,
						Version:        pkg.Version,
						Severity:       "medium",
						Description:    "Package has not been updated in over 5 years",
						Recommendation: "Check for maintained alternatives or security advisories",
					})
				}
			}
		}
	}

	summary := SecuritySummary{}
	riskLevel := "low"

	for _, vuln := range vulnerabilities {
		switch vuln.Severity {
		case "critical":
			summary.Critical++
		case "high":
			summary.High++
		case "medium":
			summary.Medium++
		case "low":
			summary.Low++
		}
	}

	if summary.Critical > 0 {
		riskLevel = "critical"
	} else if summary.High > 0 {
		riskLevel = "high"
	} else if summary.Medium > 0 {
		riskLevel = "medium"
	}

	result := SecurityAuditResult{
		Vulnerabilities: vulnerabilities,
		RiskLevel:       riskLevel,
		Summary:         summary,
	}

	jsonData, err := json.MarshalIndent(result, "", "  ")
	if err != nil {
		return "", err
	}

	return string(jsonData), nil
}

// LicenseAnalysisResult represents license analysis output
type LicenseAnalysisResult struct {
	Distribution        []types.LicenseDistribution `json:"distribution"`
	CompatibilityIssues []string                    `json:"compatibilityIssues"`
	Summary             LicenseSummary              `json:"summary"`
}

// LicenseSummary represents license statistics
type LicenseSummary struct {
	TotalPackages    int `json:"totalPackages"`
	UniqueLicenses   int `json:"uniqueLicenses"`
	UnknownLicenses  int `json:"unknownLicenses"`
}

// AnalyzeLicenses analyzes license distribution and compatibility
func AnalyzeLicenses(repoPath string) (string, error) {
	lock, err := composer.ReadComposerLock(repoPath)
	if err != nil {
		return "", err
	}

	licenseMap := make(map[string][]string)
	unknownCount := 0

	allPackages := append([]types.PackageInfo{}, lock.Packages...)
	if lock.PackagesDev != nil {
		allPackages = append(allPackages, lock.PackagesDev...)
	}

	for _, pkg := range allPackages {
		licenses := pkg.License
		if len(licenses) == 0 {
			licenses = []string{"Unknown"}
		}

		for _, license := range licenses {
			if license == "Unknown" {
				unknownCount++
			}
			licenseMap[license] = append(licenseMap[license], pkg.Name)
		}
	}

	distribution := make([]types.LicenseDistribution, 0, len(licenseMap))
	for license, packages := range licenseMap {
		distribution = append(distribution, types.LicenseDistribution{
			License:   license,
			Count:     len(packages),
			Packages:  packages,
			RiskLevel: assessLicenseRisk(license),
		})
	}

	// Check for compatibility issues
	compatibilityIssues := make([]string, 0)
	hasGPL := false
	hasProprietary := false

	for _, dist := range distribution {
		if strings.Contains(dist.License, "GPL") {
			hasGPL = true
		}
		if strings.Contains(dist.License, "Proprietary") {
			hasProprietary = true
		}
	}

	if hasGPL && hasProprietary {
		compatibilityIssues = append(compatibilityIssues,
			"Potential conflict: GPL and Proprietary licenses detected. Review compatibility.")
	}

	result := LicenseAnalysisResult{
		Distribution:        distribution,
		CompatibilityIssues: compatibilityIssues,
		Summary: LicenseSummary{
			TotalPackages:   len(allPackages),
			UniqueLicenses:  len(licenseMap),
			UnknownLicenses: unknownCount,
		},
	}

	jsonData, err := json.MarshalIndent(result, "", "  ")
	if err != nil {
		return "", err
	}

	return string(jsonData), nil
}

// assessLicenseRisk determines risk level for a license
func assessLicenseRisk(license string) string {
	safeLicenses := []string{"MIT", "Apache-2.0", "BSD-3-Clause", "BSD-2-Clause", "ISC"}
	for _, safe := range safeLicenses {
		if license == safe {
			return "safe"
		}
	}

	cautionLicenses := []string{"LGPL", "MPL", "EPL"}
	for _, caution := range cautionLicenses {
		if strings.Contains(license, caution) {
			return "caution"
		}
	}

	if strings.Contains(license, "GPL") || license == "Unknown" || strings.Contains(license, "Proprietary") {
		return "review-required"
	}

	return "caution"
}

File: dpb-mcp-rust/src/analyzer/security.rs
use anyhow::Result;
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::path::Path;

use crate::composer::read_composer_lock;
use crate::types::{LicenseDistribution, SecurityVulnerability};

#[derive(Debug, Serialize, Deserialize)]
pub struct SecurityAuditResult {
    pub vulnerabilities: Vec<SecurityVulnerability>,
    #[serde(rename = "riskLevel")]
    pub risk_level: String,
    pub summary: SecuritySummary,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct SecuritySummary {
    pub critical: usize,
    pub high: usize,
    pub medium: usize,
    pub low: usize,
}

pub fn audit_security<P: AsRef<Path>>(repo_path: P) -> Result<String> {
    let lock = read_composer_lock(&repo_path)?;

    let mut vulnerabilities = Vec::new();

    let mut all_packages = lock.packages.clone();
    if let Some(dev_packages) = &lock.packages_dev {
        all_packages.extend(dev_packages.clone());
    }

    for pkg in &all_packages {
        // Check for dev versions
        if pkg.version.contains("dev") && !pkg.version.starts_with("dev-") {
            vulnerabilities.push(SecurityVulnerability {
                package: pkg.name.clone(),
                version: pkg.version.clone(),
                severity: "medium".to_string(),
                cve: None,
                description: "Using development version in production".to_string(),
                recommendation: "Pin to a stable release version".to_string(),
            });
        }

        // Check for pre-1.0 versions
        if pkg.version.starts_with("0.") {
            vulnerabilities.push(SecurityVulnerability {
                package: pkg.name.clone(),
                version: pkg.version.clone(),
                severity: "low".to_string(),
                cve: None,
                description: "Using pre-1.0 version (potentially unstable)".to_string(),
                recommendation: "Consider upgrading to a stable 1.x+ version if available"
                    .to_string(),
            });
        }

        // Check for very old packages (5+ years)
        if let Some(time_str) = &pkg.time {
            if let Ok(pkg_time) = time_str.parse::<DateTime<Utc>>() {
                let five_years_ago = Utc::now() - chrono::Duration::days(5 * 365);
                if pkg_time < five_years_ago {
                    vulnerabilities.push(SecurityVulnerability {
                        package: pkg.name.clone(),
                        version: pkg.version.clone(),
                        severity: "medium".to_string(),
                        cve: None,
                        description: "Package has not been updated in over 5 years".to_string(),
                        recommendation: "Check for maintained alternatives or security advisories"
                            .to_string(),
                    });
                }
            }
        }
    }

    let mut summary = SecuritySummary {
        critical: 0,
        high: 0,
        medium: 0,
        low: 0,
    };

    let mut risk_level = "low";

    for vuln in &vulnerabilities {
        match vuln.severity.as_str() {
            "critical" => {
                summary.critical += 1;
                risk_level = "critical";
            }
            "high" => {
                summary.high += 1;
                if risk_level != "critical" {
                    risk_level = "high";
                }
            }
            "medium" => {
                summary.medium += 1;
                if risk_level != "critical" && risk_level != "high" {
                    risk_level = "medium";
                }
            }
            "low" => summary.low += 1,
            _ => {}
        }
    }

    let result = SecurityAuditResult {
        vulnerabilities,
        risk_level: risk_level.to_string(),
        summary,
    };

    Ok(serde_json::to_string_pretty(&result)?)
}

#[derive(Debug, Serialize, Deserialize)]
pub struct LicenseAnalysisResult {
    pub distribution: Vec<LicenseDistribution>,
    #[serde(rename = "compatibilityIssues")]
    pub compatibility_issues: Vec<String>,
    pub summary: LicenseSummary,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct LicenseSummary {
    #[serde(rename = "totalPackages")]
    pub total_packages: usize,
    #[serde(rename = "uniqueLicenses")]
    pub unique_licenses: usize,
    #[serde(rename = "unknownLicenses")]
    pub unknown_licenses: usize,
}

pub fn analyze_licenses<P: AsRef<Path>>(repo_path: P) -> Result<String> {
    let lock = read_composer_lock(&repo_path)?;

    let mut license_map: HashMap<String, Vec<String>> = HashMap::new();
    let mut unknown_count = 0;

    let mut all_packages = lock.packages.clone();
    if let Some(dev_packages) = &lock.packages_dev {
        all_packages.extend(dev_packages.clone());
    }

    for pkg in &all_packages {
        let licenses = pkg.license.clone().unwrap_or_else(|| vec!["Unknown".to_string()]);

        for license in licenses {
            if license == "Unknown" {
                unknown_count += 1;
            }
            license_map
                .entry(license)
                .or_insert_with(Vec::new)
                .push(pkg.name.clone());
        }
    }

    let unique_license_count = license_map.len();
    
    let distribution: Vec<LicenseDistribution> = license_map
        .into_iter()
        .map(|(license, packages)| LicenseDistribution {
            risk_level: assess_license_risk(&license),
            count: packages.len(),
            license,
            packages,
        })
        .collect();

    // Check for compatibility issues
    let mut compatibility_issues = Vec::new();
    let has_gpl = distribution.iter().any(|d| d.license.contains("GPL"));
    let has_proprietary = distribution.iter().any(|d| d.license.contains("Proprietary"));

    if has_gpl && has_proprietary {
        compatibility_issues.push(
            "Potential conflict: GPL and Proprietary licenses detected. Review compatibility."
                .to_string(),
        );
    }

    let result = LicenseAnalysisResult {
        distribution,
        compatibility_issues,
        summary: LicenseSummary {
            total_packages: all_packages.len(),
            unique_licenses: unique_license_count,
            unknown_licenses: unknown_count,
        },
    };

    Ok(serde_json::to_string_pretty(&result)?)
}

fn assess_license_risk(license: &str) -> String {
    let safe_licenses = ["MIT", "Apache-2.0", "BSD-3-Clause", "BSD-2-Clause", "ISC"];

    if safe_licenses.contains(&license) {
        return "safe".to_string();
    }

    let caution_licenses = ["LGPL", "MPL", "EPL"];
    for caution in &caution_licenses {
        if license.contains(caution) {
            return "caution".to_string();
        }
    }

    if license.contains("GPL") || license == "Unknown" || license.contains("Proprietary") {
        return "review-required".to_string();
    }

    "caution".to_string()
}
